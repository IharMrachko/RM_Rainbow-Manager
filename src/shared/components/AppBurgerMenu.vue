<template>
  <div data-burger-test-id class="burger" :class="{ open: isOpen }" @click="toggle">
    <span></span>
    <span></span>
    <span></span>
  </div>
</template>
<script lang="ts">
import { defineComponent, ref } from 'vue';
import { useStore } from 'vuex';

export default defineComponent({
  setup() {
    const store = useStore();
    const isOpen = ref(false);
    const toggle = () => {
      isOpen.value = !isOpen.value;
      store.dispatch('burgerMenu/setBurger', { isOpen: isOpen.value });
    };
    return {
      isOpen,
      toggle,
    };
  },
});
</script>
<style scoped lang="scss">
.burger {
  position: relative;
  width: 30px;
  height: 24px;
  cursor: pointer;
}

.burger span {
  position: absolute;
  left: 0;
  width: 100%;
  height: 4px;
  background: #fff;
  border-radius: 2px;
  transition: 0.3s;
}

/* три линии */
.burger span:nth-child(1) {
  top: 0;
}
.burger span:nth-child(2) {
  top: 10px;
}
.burger span:nth-child(3) {
  top: 20px;
}

/* анимация в крест */
.burger.open span:nth-child(1) {
  transform: rotate(45deg);
  top: 10px;
}
.burger.open span:nth-child(2) {
  opacity: 0;
}
.burger.open span:nth-child(3) {
  transform: rotate(-45deg);
  top: 10px;
}
</style>
