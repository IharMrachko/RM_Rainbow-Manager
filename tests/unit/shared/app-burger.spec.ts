import { mount } from '@vue/test-utils';
import AppBurgerMenu from '@/shared/components/AppBurgerMenu.vue';
import { createStore } from 'vuex';

describe('AppBurger', () => {
  let wrapper: any;
  let setBurgerMock: jest.Mock;

  beforeEach(() => {
    setBurgerMock = jest.fn((ctx, payload) => {
      // имитируем мутацию стора
      ctx.state.isOpen = payload.isOpen;
    });

    const store = createStore({
      modules: {
        burgerMenu: {
          namespaced: true,
          state: () => ({
            isOpen: false,
          }),
          getters: {
            isOpen: (state) => state.isOpen,
          },
          actions: {
            setBurger: setBurgerMock,
          },
        },
      },
    });

    wrapper = mount(AppBurgerMenu, {
      global: {
        plugins: [store],
      },
    });
  });

  afterEach(() => {
    wrapper.unmount();
    wrapper = null;
  });

  it('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('if burger click should burger contain class open', async () => {
    const burger = wrapper.find('[data-burger-test-id]');
    await burger.trigger('click');
    expect(burger.classes()).toContain('open');
  });

  it('dispatches correct action with isOpen true/false', async () => {
    const burger = wrapper.find('[data-burger-test-id]');
    await burger.trigger('click');
    expect(setBurgerMock).toHaveBeenCalledWith(expect.any(Object), {
      isOpen: true,
    });
    await burger.trigger('click');
    expect(setBurgerMock).toHaveBeenCalledWith(expect.any(Object), {
      isOpen: false,
    });
  });
  it('renders exactly 3 span lines', () => {
    const spans = wrapper.findAll('span');
    expect(spans).toHaveLength(3);
  });

  it('applies open class styles to spans', async () => {
    const burger = wrapper.find('[data-burger-test-id]');
    await burger.trigger('click');
    expect(burger.classes()).toContain('open');
    // можно проверить конкретный span
    const spans = burger.findAll('span');
    expect(spans[1].element.getAttribute('style')).toBeNull(); // стили применяются через CSS, но класс есть
  });
});
