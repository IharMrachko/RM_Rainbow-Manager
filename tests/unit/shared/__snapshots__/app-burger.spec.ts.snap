// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppBurger matches snapshot 1`] = `<div data-burger-test-id="" class="burger"><span></span><span></span><span></span></div>`;
